File: apps/server/openhive/events/schema.py
"""Typed event schema. Every engine step emits one or more of these.

The same events are persisted to SQLite `run_events` AND fanned out via SSE.
The Run-mode canvas and the Timeline tab both read from this stream.
"""

from __future__ import annotations

import time
from typing import Any, Literal

from pydantic import BaseModel, Field


EventKind = Literal[
    "run_started",
    "run_finished",
    "run_error",
    "node_started",
    "node_finished",
    "token",
    "tool_called",
    "tool_result",
    "delegation_opened",
    "delegation_closed",
    "checkpoint",
]


class Event(BaseModel):
    """One envelope for every event. Unused fields are simply omitted."""

    kind: EventKind
    ts: float = Field(default_factory=time.time)
    run_id: str
    # depth 0 = top-level run. Delegated sub-runs increment depth.
    depth: int = 0
    node_id: str | None = None
    # tool / delegation context
    tool_call_id: str | None = None
    tool_name: str | None = None
    # free-form payload — the UI picks fields by kind
    data: dict[str, Any] = Field(default_factory=dict)


def make_event(
    kind: EventKind,
    run_id: str,
    *,
    depth: int = 0,
    node_id: str | None = None,
    tool_call_id: str | None = None,
    tool_name: str | None = None,
    **data: Any,
) -> Event:
    return Event(
        kind=kind,
        run_id=run_id,
        depth=depth,
        node_id=node_id,
        tool_call_id=tool_call_id,
        tool_name=tool_name,
        data=data,
    )

File: apps/web/lib/server/events/schema.ts
/**
 * Typed event schema. Every engine step emits one or more of these.
 * Ports apps/server/openhive/events/schema.py.
 *
 * Events are persisted to SQLite `session_events` AND fanned out via SSE. The
 * Run-mode canvas and the Timeline tab read from the same stream — no side
 * channels.
 */

export type EventKind =
  | 'run_queued'
  | 'run_started'
  | 'run_finished'
  | 'run_error'
  | 'node_started'
  | 'node_finished'
  | 'node_error'
  | 'token'
  | 'tool_called'
  | 'tool_result'
  | 'delegation_opened'
  | 'delegation_closed'
  | 'checkpoint'
  | 'user_question'
  | 'user_answered'
  | 'user_message'
  | 'turn_finished'
  | 'todos_changed'
  | 'skill.queued'
  | 'skill.started'
  | 'skill.progress'
  | 'tool_run.partitioned'
  // A4 — token estimation + window math
  | 'token.estimate.drift'
  | 'turn.blocked'
  | 'turn.round_limit'
  | 'autocompact.disabled'
  // Round-level observability (multi-round tool-calling turns)
  | 'round_started'
  | 'round_finished'
  | 'provider.empty_round'
  // A2 — lifecycle hooks
  | 'hook.invoked'
  // S3 — parallel fork
  | 'fork.spawned'
  | 'fork.skipped'
  // S2 — microcompact
  | 'microcompact.applied'
  // A3 — artifact rehydration
  | 'artifact.read'
  | 'artifact.read.denied'
  // Provider-side hosted tool lifecycle (Codex web_search, Anthropic
  // server_tool_use). Emitted per phase so the UI can show "🔍 web_search
  // searching…" instead of a blank turn while a 30-150s native search runs.
  | 'native_tool'

export interface Event {
  kind: EventKind
  ts: number
  session_id: string
  depth: number
  node_id: string | null
  tool_call_id: string | null
  tool_name: string | null
  data: Record<string, unknown>
}

export interface MakeEventOpts {
  depth?: number
  node_id?: string | null
  tool_call_id?: string | null
  tool_name?: string | null
}

export function makeEvent(
  kind: EventKind,
  sessionId: string,
  data: Record<string, unknown>,
  opts: MakeEventOpts = {},
): Event {
  return {
    kind,
    ts: Date.now() / 1000,
    session_id: sessionId,
    depth: opts.depth ?? 0,
    node_id: opts.node_id ?? null,
    tool_call_id: opts.tool_call_id ?? null,
    tool_name: opts.tool_name ?? null,
    data,
  }
}
